RedundantSigningWorkarounds.targets

```diff
@@ -0,0 +1,81 @@
+  2. When signing files, drop a temp file next to it that tells us the file has been signed. If that temp file alread exists, prevent re-signing the file.
+      See targets: AddSignedIndicatorFile, RemoveRedundantSigning, AddSignedIndicatorFileForNuget, RemoveRedundantSigningForNuget
+
+  Workaround #2 could be done differently for build output by disabling signing altogether until CoreCompile runs, but that breaks workaround #1, hence the temp file approach.
+
+-->
+
+  <PropertyGroup>
+    <SigningCompleteFile>$(OutDir)\signing-completed.txt</SigningCompleteFile>
+    
+    <TargetsTriggeredByCompilation>
+        $(TargetsTriggeredByCompilation);
+        ClearSigningTempFileIfCompileRuns
+    </TargetsTriggeredByCompilation>
+  </PropertyGroup>
+
+  <!--
+    RemoveRedundantSigning and AddSignedIndicatorFile ensure we don't sign build output multiple times.
+    Once signing is complete, we add a temp file. If the temp file is already present, we skip signing.
+    -->
+  <Target Name="RemoveRedundantSigning" BeforeTargets="SignFiles" Condition="Exists($(SigningCompleteFile))">
+    <Message Text="Removing all FilesToSign, signing has already happened" Importance="High" />
+    <ItemGroup>
+      <FilesToSign Remove="@(FilesToSign)" />
+    </ItemGroup>
+  </Target>
+
+  <Target Name="AddSignedIndicatorFile" AfterTargets="SignFiles">
+      <WriteLinesToFile File="$(SigningCompleteFile)" Lines="@(FilesToSign)"/>
+  </Target>
+
+  <!-- Ensure SignNugetFiles will run before CoreCompile. -->
+  <Target Name="EnforceNugetSigningBeforeOutputSigning" Condition="'@(NuGetFilesToSign)' != ''" BeforeTargets="CoreCompile" DependsOnTargets="SignNugetFiles">
+      <Message Text="Enforing NuGet signing before core compile..." />
+  </Target>
+
+  <!-- Do not sign NuGet files that have already been signed. -->
+  <Target Name="RemoveRedundantSigningForNuget" BeforeTargets="SignNugetFiles">
+    <Message Text="Removing NugetFilesToSign where signing has already happened" />
+    <ItemGroup>
+      <NuGetFilesToSign Remove="@(NuGetFilesToSign)" Condition="Exists('%(NuGetFilesToSign.Identity).txt')" />
+    </ItemGroup>
+  </Target>
+
+  <!-- Create a temp file next to each Nuget dll, which indicates that we already signed the file -->
+  <Target Name="AddSignedIndicatorFileForNuget" AfterTargets="SignNuGetFiles">
+    <WriteLinesToFile File="%(NuGetFilesToSign.Identity).txt" Lines="NuGet Signed" Condition="'@(NuGetFilesToSign)' != ''"/>
+  </Target>
+
+  <!--
+    If CoreCompile runs, we need to clear the output dll signing file since the dll will get overwritten with an unsigned file.
+    We do *not* need to clear the nuget signing files, though. Those dlls should not get overwritten.
+  -->
+  <Target Name="ClearSigningTempFileIfCompileRuns">
+    <Message Text="Removing signing complete indicator file because CoreCompile ran..." />
+    <Delete Files="$(SigningCompleteFile)" />
+  </Target>
+</Project>
```

Comment:
You're right - we actually don't need the NuGet stuff here, at least not at the moment... (Here I was, thinking I'd just pick a submodule to send the PR as an example, and I picked the one that doesn't have any nugets :) Whoops!). I'll remove that.

We do still need a workaround to ensure the output assemblies don't get re-signed if compile isn't run, so those parts will have to stay.

Regarding the suggestions, I believe that would go against the current guidance from VSEng. At the moment, the recommendation is to sign in the specific project that produces the assembly, and if it's in a submodule, to sign in the submodule itself rather than in the repo that includes it (https://devdiv.visualstudio.com/DevDiv/_wiki/wikis/DevDiv.wiki/27099/Guidance-for-Xamarin-MicroBuild-Signing-Migration?anchor=handling-submodules).

We can have a conversation w/ VSEng about whether that's the right approach (there's already a thread in the VSEng teams channel), but so far that has not amounted to much...

Reply:
Thanks I missed this conversation in Teams, lets continue the discussion on how to best move forward there and on https://github.com/xamarin/androidtools/issues/331